drivers/usb/udc/udc_stm32.c

```diff
@@ -41,27 +41,77 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define UDC_STM32_IRQ		DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, irq)
 #define UDC_STM32_IRQ_PRI	DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, priority)
 
-#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
-			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+/* Shorthand to obtain PHY node for an instance */
+#define UDC_PHY(node_id)			DT_PROP_BY_IDX(node_id, phys, 0)
 
-#define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
-			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+/** Evaluates to 1 if PHY of @p usb_node is an embedded PHY, 0 otherwise. */
+#define UDC_STM32_PHY_IS_EMBEDDED(usb_node)	UTIL_OR(UTIL_OR(		\
+	DT_NODE_HAS_COMPAT(UDC_PHY(usb_node), st_stm32_usbphyc),		\
+	DT_NODE_HAS_COMPAT(UDC_PHY(usb_node), st_stm32u5_otghs_phy)),		\
+	DT_NODE_HAS_COMPAT(UDC_PHY(usb_node), st_stm32_embedded_usb_phy))
 
 /**
- * The following defines are used to map the value of the "maxiumum-speed"
- * DT property to the corresponding definition used by the STM32 HAL.
+ * Evaluates to 1 if PHY of @p usb_node is an embedded HS PHY, 0 otherwise.
+ *
+ * @note Undefined behavior if UDC_PHY_IS_EMBEDDED(usb_node) is not 1.
  */
-#if defined(CONFIG_SOC_SERIES_STM32H7X) || USB_OTG_HS_EMB_PHY
-#define UDC_STM32_HIGH_SPEED             USB_OTG_SPEED_HIGH_IN_FULL
-#else
-#define UDC_STM32_HIGH_SPEED             USB_OTG_SPEED_HIGH
-#endif
+#define UDC_STM32_EMBEDDED_PHY_IS_HS(usb_node)	UTIL_OR(UTIL_OR(		\
+	DT_NODE_HAS_COMPAT(UDC_PHY(usb_node), st_stm32_usbphyc),		\
+	DT_NODE_HAS_COMPAT(UDC_PHY(usb_node), st_stm32u5_otghs_phy)),		\
+	DT_PROP(UDC_PHY(usb_node), high_speed))
 
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usb)
-#define UDC_STM32_FULL_SPEED             PCD_SPEED_FULL
-#else
-#define UDC_STM32_FULL_SPEED             USB_OTG_SPEED_FULL
-#endif
+/** Evaluates to 1 if @p inst can run in High-Speed mode, 0 otherwise. */
+#define UDC_STM32_NODE_HS_CAPABLE(node_id)	DT_NODE_HAS_COMPAT(node_id, st_stm32_otghs)
+
+/**
+ * Returns the 'PCD_PHY_Module' value for @p node_id
+ *
+ * This value may be one of:
+ *    - PCD_PHY_EMBEDDED: embedded Full-Speed PHY
+ *    - PCD_PHY_UTMI: embedded High-Speed PHY over UTMI+
+ *    - PCD_PHY_ULPI: external High-Speed PHY over ULPI
+ *
+ * The PHY DT node referenced by the 'phys' property of @p node_id
+ * is examined to determine the correct value:
+ *  - compatible 'st,stm32-usbphyc'    => embedded HS PHY
```

Comment:
Wouldn't it clarify things to rename `st,stm32-usbphyc` to `st,stm32-usb-hs-phyc` ?